codalab/lib/upload_manager.py

```diff
@@ -1,74 +1,113 @@
+        Args:
+            source_ext (str): File extension of the source to write.
+            source_fileobj (str): Fileobj of the source to write.
+            bundle_path (str): Output bundle path.
+            unpack_archive (bool): Whether fileobj is an archive that should be unpacked.
         """
-        bundle_path = self._bundle_store.get_bundle_location(bundle.uuid)
+        raise NotImplementedError
+
+    def upload_to_bundle_store(self, bundle: Bundle, source: Source, git: bool, unpack: bool):
+        """Uploads the given source to the bundle store.
+        Given arguments are the same as UploadManager.upload_to_bundle_store()"""
         try:
+            # bundle_path = self._bundle_store.get_bundle_location(bundle.uuid)
             is_url, is_fileobj, filename = self._interpret_source(source)
             if is_url:
                 assert isinstance(source, str)
                 if git:
-                    file_util.git_clone(source, bundle_path)
+                    bundle_path = self._update_and_get_bundle_location(bundle, is_directory=True)
+                    self.write_git_repo(source, bundle_path)
                 else:
                     # If downloading from a URL, convert the source to a file object.
                     is_fileobj = True
                     source = (filename, urlopen_with_retry(source))
             if is_fileobj:
-                if unpack and self.zip_util.path_is_archive(filename):
-                    self._unpack_fileobj(source[0], source[1], bundle_path)
+                source_filename, source_fileobj = cast(Tuple[str, IO[bytes]], source)
+                source_ext = zip_util.get_archive_ext(source_filename)
+                if unpack and zip_util.path_is_archive(filename):
+                    bundle_path = self._update_and_get_bundle_location(
+                        bundle, is_directory=source_ext in ARCHIVE_EXTS_DIR
+                    )
+                    self.write_fileobj(source_ext, source_fileobj, bundle_path, unpack_archive=True)
                 else:
-                    with open(bundle_path, 'wb') as out:
-                        shutil.copyfileobj(cast(IO, source[1]), out)
+                    bundle_path = self._update_and_get_bundle_location(bundle, is_directory=False)
+                    self.write_fileobj(
+                        source_ext, source_fileobj, bundle_path, unpack_archive=False
+                    )
 
-            # is_directory is True if the bundle is a directory and False if it is a single file.
-            is_directory = os.path.isdir(bundle_path)
-            self._bundle_model.update_bundle(
-                bundle, {'storage_type': StorageType.DISK_STORAGE.value, 'is_dir': is_directory},
-            )
         except UsageError:
             if os.path.exists(bundle_path):
                 path_util.remove(bundle_path)
             raise
 
+    def _update_and_get_bundle_location(self, bundle: Bundle, is_directory: bool) -> str:
```

Comment:
why did we combine this functionality (update & get)? would it be better for the code to be more granular?

Reply:
Because they're only ever called together. Additionally, the "update" part of the code affects the result of "getting" the bundle location (get_bundle_location constructs a bundle location based on the storage_type and is_directory fields).